docs: rewrite the traffic shaping (#2353)
apollographql/router: docs/source/configuration/traffic-shaping.mdx

title: Traffic shaping in the Apollo Router
---

The Apollo Router provides various features to improve the performance and reliability of

> the traffic with clients and subgraphs.

Reviewer: ```suggestion
the traffic between clients and subgraphs.
```

Author: This is about the traffic between client and router, and between router and subgraph

Reviewer: Yep. I was trying to be more clear, but perhaps we just need to be more explicit, e.g.:
```
the traffic between the client and router and between the router and subgraphs.
```
?

Author: fixed in 4e3f567b

        ttl: 10s # for each successful request, we register a token, that expires according to this option (default: 10s)
        retry_percent: 0.2 # defines the proportion of available retries to the current number of tokens
        retry_mutations: false # allows retries on mutations. This should only be enabled if mutations are idempotent
```

## Client side traffic shaping

### Rate limiting

The Apollo Router can apply rate limiting on client requests, as follows:

```yaml title="router.yaml"
traffic_shaping:
  router: # Rules applied to requests from clients to the router
    global_rate_limit: # Accept a maximum of 10 requests per 5 secs. Excess requests must be rejected.
      capacity: 10
      interval: 5s # Must not be greater than 18_446_744_073_709_551_615 milliseconds and not less than 0 milliseconds
```

> This rate limiting applies to all requests, there is no filtering per IP or other criterion.

Reviewer: ```suggestion
This rate limiting applies to all requests, there is no filtering per IP or other criteria.
```

### Timeout

The Apollo Router applies a default limit of 30 seconds to receive the entire client request. That limit is configurable:

```yaml title="router.yaml"
traffic_shaping:
  router: 
    timeout: 50s # If a request to the router takes more than 50secs then cancel the request (30 sec by default)
```

### Compression

Compression is automatically supported on the client side, depending on the `Accept-Encoding` header provided by the client.

> ## Subgraph side traffic shaping

Reviewer 2: Would removing the word "side" here still provide the right precision?

```suggestion
## Subgraph traffic shaping
```

Otherwise I can read this as "Subgraph side-traffic shaping", too.

> The Apollo Router supports various options affecting traffic with subgraphs, that can either be defined for all subgraphs,

Reviewer 2: ```suggestion
The Apollo Router supports various options affecting traffic destined for subgraphs, that can either be defined for all subgraphs,
```

or overriden per subgraph:

```yaml title="router.yaml"
traffic_shaping:
  all:
    deduplicate_query: true # Enable query deduplication for all subgraphs.
  subgraphs: # Rules applied to requests from the router to individual subgraphs
    products:
      deduplicate_query: false # Disable query deduplication for the products subgraph.
```

### Compression

The Apollo Router can compress request bodies to subgraphs (along with response bodies to clients).
It currently supports these algorithms: `gzip`, `br`, and `deflate`.

```yaml title="router.yaml"
traffic_shaping:
  all:

>     compression: br # Enable brotli compression for all subgraphs.

Reviewer 2: Is anything enabled by default here?

Author: it's not enabled by default, but now I see that decompressing the subgraph response will only work if we compress the subgraph request https://github.com/apollographql/router/blob/dev/apollo-router/src/plugins/traffic_shaping/mod.rs#L353-L357
I think we should always accept compressed subgraph responses,  right?

Author: opening https://github.com/apollographql/router/issues/2415 about this

```

### Rate limiting

Subgraph request rate limiting uses the same configuration as client rate limiting, and is calculated
per subgraph, not per backend host.

```yaml title="router.yaml"
traffic_shaping:
  all:
    global_rate_limit: # Accept a maximum of 10 requests per 5 secs. Excess requests must be rejected.
      capacity: 10
      interval: 5s # Must not be greater than 18_446_744_073_709_551_615 milliseconds and not less than 0 milliseconds
```

### Experimental request retry

On failure, subgraph requests can be retried automatically. This is deactivated by default for mutations.

> This uses [Finagle's *RetryBudget* algorithm](https://finagle.github.io/blog/2016/02/08/retry-budgets/),

Reviewer 2: Fwiw, this isn't about this line specifically, but I believe the documentation team's stylistic preference is to not have hard-wraps.

Author: I updated the harsd wraps in [f3425c8](https://github.com/apollographql/router/pull/2353/commits/f3425c885467d36e78a5af232dccdd563312beb0)

in which every successful request adds an expirable token to a bucket, and every retry consumes a number
of those tokens. On top of that, a minimal number of retries per second is available, to test regularly
when the retry budget was entirely consumed or on startup when very few requests have been sent.
The tokens expire so the budget has a large number of available retries if a lot of recent requests were

> successful, but reduces quickly on frequent failures to avoid sending too much traffic to the subgraph.

Reviewer 2: ```suggestion
successful but reduces quickly on frequent failures to avoid sending too much traffic to the subgraph.
```

It is configurable as follows:

```yaml title="router.yaml"
traffic_shaping:
  all:
    experimental_retry:
      min_per_sec: 10 # minimal number of retries per second (`min_per_sec`, default is 10 retries per second)
      ttl: 10s # for each successful request, we register a token, that expires according to this option (default: 10s)
      retry_percent: 0.2 # defines the proportion of available retries to the current number of tokens
      retry_mutations: false # allows retries on mutations. This should only be enabled if mutations are idempotent
```

> ### Variable deduplication

Reviewer: Not this PR, but...

Why is this configurable? Why wouldn't you always want to do this? Same for Query deduplication...

I vaguely remember there was a reason for this at one time, but I can't remember the reason and I wonder if it's no longer relevant.

Author: IIRC The reason was mainly that we don't know if it's stable enough, so it should have an option to deactivate it. I agree that it should always be active

Reviewer 2: I think variable de-duplication is probably fine.

> Query deduplication

This does tend to have more side effects (behind the scenes things), generally speaking, which is why we initially avoided it.

Author: I opened https://github.com/apollographql/router/issues/2387 to keep track of that

> In federated queries, using the `_entities` operation, it is often the case that one entity is requested from
> multiple parts of a query. As an example, for each review of each product, geting the author's name might return
> the same name multiple times if an author wrote multiple reviews.
> 

Reviewer 2: ```suggestion
When subgraphs are sent entity requests by the Router using the `_entities` field, it is often the case that the same entity (identified by a unique `@key` constraint) is requested multiple times within the execution of a single federated query.  For example, an author's name might need to be fetched multiple times when accessing a list of a reviews for a product for which the author has written multiple reviews.

```

To reduce the size of subgraph requests and the amount of work they might perform, the list of entities sent
can be deduplicated. This is activated with the `deduplicate_variables` option:

```yaml title="router.yaml"
traffic_shaping:
  deduplicate_variables: true # Enable the variable deduplication optimization.
```

### Query deduplication

> If similar queries are sent to a subgraph, the Apollo Router can avoid sending them multiple times, and instead
> copy the result of the first query when it comes back from the subgraph, thus reducing the traffic to the subgraph
> and the overall client request latency.
> 
> The subgraph queries have to be exactly the same (path, headers and body) to be deduplicated.
> 

Reviewer 2: ```suggestion
If the Router is simultaneously processing similar queries, it may result in producing multiple identical requests to a subgraph.  With the `deduplicate_query` functionality enabled, the Router can avoid sending the same query multiple times and instead buffer one or more of the dependent queries pending the result of the first, and reuse that result to fulfill all of the initial queries.  This will reduce the overall traffic to the subgraph and the overall client request latency.  To meet the criteria for deduplication, the feature must be enabled and the subgraph queries must have have the same HTTP path, headers and body:

```

```yaml title="router.yaml"
traffic_shaping:
  all:
    deduplicate_query: true # Enable query deduplication for all subgraphs.
```

> ### Ordering

Reviewer: If this means "ordering in the configuration", then this is really important and should be right at the top of the page with a little more text to explain why.

If it doesn't mean that, then what's the purpose of this section?

Author: This is to explain in which order the different steps are applied, so we can understand how they interact. The configuration order will have no impact on it.

Reviewer: Ok. In that case, perhaps re-word that sentence to be more clear about why we are telling people this. Something like:

```
Traffic shaping will alway be executed in this order within the router in order to ensure that interactions between different steps are correct. Declaration order in configuration does not affect this ordering.
```

Author: fixed in 4e3f567b
